File: web/src/types/film.ts
/**
 * 影视 DTO —— 对齐新后端 /api/v1 契约(server/internal/dto/film.go + openapi.yaml)。
 * 关键: cover(非 picture)、sources/episodes(非 list/linkList)、related(非 relate)、
 *       dbScore:number、mid 主键、详情字段扁平(无 descriptor)。
 */

/** 导航分类(带子分类) */
export interface NavCategory {
  id: number
  pid: number
  name: string
  children?: NavCategory[]
}

/** 影片卡片(列表/首页/搜索/推荐) */
export interface Card {
  mid: number
  name: string
  cover: string
  /** 横版大图(16:9, 轮播/横幅背景用)。来自 TMDB 回填(movie_search.backdrop), 后端按
   *  "有则返回、无则字段缺省"输出; 一律按"有则用、无则回退 cover"处理。 */
  poster?: string
  cid: number
  pid: number
  cName: string
  subTitle: string
  area: string
  year: number
  state: string
  remarks: string
  dbScore: number
  /** 上映日期(ISO 前缀串: 2026-09-11 / 2026-07 / 2007)。源站未提供时字段缺省 */
  pubDate?: string
  /** 当前豆瓣榜位(1 起)。不在榜时字段缺省 —— 卡片角标 "Hot No.N" 用 */
  hotRank?: number
  /** 榜位来源榜单中文名(如 "热门电影")。不在榜时缺省 —— 无轮播位时这条卡片要兜底当首屏大图,
   *  描述行靠它把 "No.N" 说清是哪个榜 */
  hotBoard?: string
  /** 类型标签(如 "动作,冒险")。同上, 兜底首屏大图时拆成标签行; 无标签影片缺省 */
  classTag?: string
}

/** 首页轮播项(GET /banners) —— 后端生效位: 手动配置位 + 热榜自动补位, 与后台管理页同源。 */
export interface HomeBanner {
  /** banner=手动配置位 / fallback=热榜自动补位 */
  source: 'banner' | 'fallback'
  mid?: number
  name: string
  subtitle?: string
  /** 横图(宽幅主视觉; 自动位为 TMDB 回填图, 可能缺省) */
  image?: string
  /** 竖图/封面 */
  poster?: string
  /** 自定义跳转(站内路径或外链); 优先于 mid */
  link?: string
  /** 影片元信息(带 mid 的生效位由后端按 mid 补齐; 纯自定义位缺省) —— 首屏大图描述行用 */
  dbScore?: number
  /** 类型标签(如 "动作,冒险"), 前端按逗号/顿号/斜杠拆分展示 */
  classTag?: string
  /** 当前豆瓣榜位(1 起), 缺省/0 = 不在榜 */
  hotRank?: number
  /** 榜位来源榜单中文名(如 "热门电影"), 展示成「豆瓣·热门电影 No.1」 */
  hotBoard?: string
}

/** HeroCarousel 单项 —— 兼容"影片卡片(Card)"与"后台 Banner"两种来源。 */
export interface HeroItem {
  mid?: number
  name: string
  /** 竖图/封面 */
  cover?: string
  /** 横图(宽幅主视觉) */
  poster?: string
  cName?: string
  area?: string
  year?: number
  remarks?: string
  /** 自定义跳转(站内路径或外链); 有则优先于 mid */
  link?: string
  /** 评分(0/缺省 = 无评分) —— 与详情页 dbScore 同源 */
  dbScore?: number
  /** 类型标签(classTag 原始串, 前端拆分) */
  classTag?: string
  /** 豆瓣榜位(0/缺省 = 不在榜) */
  hotRank?: number
  /** 榜位来源榜单中文名 */
  hotBoard?: string
}

/** 单集 */
export interface Episode {
  episode: string
  link: string
}

/** 一个播放源 */
export interface PlaySource {
  id: string
  name: string
  episodes: Episode[]
  /** 服务端 m3u8 可达性(广告过滤代理链路): false=直接走直链跳过过滤代理; undefined=未测 */
  adFilterOk?: boolean
}

/** 影片详情(字段扁平 + 多源) */
export interface FilmDetail {
  mid: number
  name: string
  cover: string
  /** 横图(16:9, 详情页 hero 背景用)。TMDB 回填后才有, 缺省为 undefined, 使用处须兜底 cover */
  backdrop?: string
  cid: number
  pid: number
  cName: string
  subTitle: string
  actor: string
  director: string
  area: string
  language: string
  year: number
  classTag: string
  remarks: string
  state: string
  dbScore: number
  /** 上映日期(ISO 前缀串, 源站未提供时缺省) */
  pubDate?: string
  /** 当前豆瓣榜位(1 起, 不在榜缺省) —— 详情页「豆瓣·热门电影 No.N」用 */
  hotRank?: number
  /** 榜位来源榜单中文名(如 "热门电影"); 同一片多集合各有位次, 展示具体是哪个榜的 No.X */
  hotBoard?: string
  content: string
  playFrom: string[]
  sources: PlaySource[]
}

/** GET /films/{mid} 响应 */
export interface FilmDetailResp {
  detail: FilmDetail
  related: Card[]
}

/** GET /films/{mid}/play 响应 */
export interface PlayInfo {
  detail: FilmDetail
  current: Episode
  currentSource: string
  currentEpisode: number
  related: Card[]
}

/** 首页区块 */
export interface HomeRow {
  nav: NavCategory
  latest: Card[]
  hot: Card[]
}

/** GET /home 响应 */
export interface HomeData {
  categories: NavCategory[]
  /** 全站跨类别热榜(「🔥 热门榜单」行) —— 与 rows[].hot(该分类热榜)口径不同, 后端已混排好 */
  hot: Card[]
  rows: HomeRow[]
}

/** GET /films/classify 响应 */
export interface ClassifyData {
  title?: NavCategory
  news: Card[]
  top: Card[]
  recent: Card[]
  /** 高分榜(db_score 降序, 排除解说) —— scoredCount 为 0 时必为空数组, 不渲染分区 */
  score: Card[]
  /** 该分类有评分的影片总数(0 = 该分类无评分数据, 隐藏高分榜入口) */
  scoredCount: number
}

/** 筛选标签(小写 name/value, 对齐后端 TagOption) */
export interface FilterTag {
  name: string
  value: string
}

/** GET /categories/{pid}/filters 响应 */
export interface Filters {
  titles: Record<string, string>
  tags: Record<string, FilterTag[]>
  sortList: string[]
}

/** GET /films 查询参数(camelCase) */
export interface FilmsQuery {
  keyword?: string
  pid?: number
  category?: number
  plot?: string
  area?: string
  language?: string
  year?: number
  sort?: string
  page?: number
  size?: number
}

File: server/internal/dto/film.go
package dto

import (
	"server/internal/douban"
	"server/internal/domain/entity"
	"server/internal/domain/repository"
	"server/internal/service"
	"server/internal/tmdb"
)

// normBackdrop 对外输出前归一化: 检索无果哨兵不外泄, 统一为空串。
func normBackdrop(b string) string {
	if b == tmdb.MissMark {
		return ""
	}
	return b
}

// Card 影片卡片(裁剪字段, 跨国线路减重)。
type Card struct {
	Mid     int64   `json:"mid"`
	Name    string  `json:"name"`
	Cover   string  `json:"cover"`
	// Poster 横图(16:9)。数据列是 movie_search.backdrop; JSON 字段名沿用前端 TS 契约的
	// poster(前端早已按"有则用、无则回退 cover"预留了该字段), 首页兜底轮播零改动点亮。
	Poster  string  `json:"poster,omitempty"`
	Cid     int64   `json:"cid"`
	Pid     int64   `json:"pid"`
	CName   string  `json:"cName"`
	SubTitle string  `json:"subTitle"`
	Area     string  `json:"area"`
	Year     int     `json:"year"`
	State    string  `json:"state"`
	Remarks  string  `json:"remarks"`
	DbScore  float64 `json:"dbScore"`
	// PubDate 上映日期(ISO 前缀串: 2026-09-11 / 2026-07 / 2007), 空 = 源站未提供。
	// 源站覆盖率低且随机, 故 omitempty —— 没值就不占 payload。
	PubDate string `json:"pubDate,omitempty"`
	// HotRank 当前豆瓣榜位(1 起), 0 = 不在榜。前端据此显示 "Hot No.N" 角标;
	// omitempty 让 15 万部榜外影片的卡片不带这个字段。
	HotRank int `json:"hotRank,omitempty"`
}

func ToCard(m entity.MovieSearch) Card {
	return Card{
		Mid: m.Mid, Name: m.Name, Cover: m.Cover, Poster: normBackdrop(m.Backdrop), Cid: m.Cid,
		Pid: m.Pid, CName: m.CName,
		SubTitle: m.SubTitle, Area: m.Area, Year: m.Year, State: m.State, Remarks: m.Remarks, DbScore: m.DbScore,
		PubDate: m.PubDate, HotRank: m.HotRank,
	}
}

func ToCards(list []entity.MovieSearch) []Card {
	out := make([]Card, 0, len(list))
	for _, m := range list {
		out = append(out, ToCard(m))
	}
	return out
}

// ManageFilmRow 后台影片列表行: 公开 Card + 软删标记。
// 单独定义而不给 Card 加字段, 是为了不把 deletedAt 泄漏到公开接口契约里。
type ManageFilmRow struct {
	Card
	DeletedAt int64 `json:"deletedAt"`
}

func ToManageFilmRow(m entity.MovieSearch) ManageFilmRow {
	return ManageFilmRow{Card: ToCard(m), DeletedAt: m.DeletedAt}
}

// ToManageFilmRows 后台影片列表批量转换。
func ToManageFilmRows(list []entity.MovieSearch) []ManageFilmRow {
	out := make([]ManageFilmRow, 0, len(list))
	for _, m := range list {
		out = append(out, ToManageFilmRow(m))
	}
	return out
}

// Episode 单集。
type Episode struct {
	Episode string `json:"episode"`
	Link    string `json:"link"`
}

// PlaySource 一个播放源(契约: id/name/episodes; adFilterOk 服务端 m3u8 可达性, false 时播放页跳过代理过滤)。
type PlaySource struct {
	Id         string    `json:"id"`
	Name       string    `json:"name"`
	Episodes   []Episode `json:"episodes"`
	AdFilterOk *bool     `json:"adFilterOk,omitempty"`
}

func toEpisodes(eps []entity.Episode) []Episode {
	out := make([]Episode, 0, len(eps))
	for _, e := range eps {
		out = append(out, Episode{Episode: e.Episode, Link: e.Link})
	}
	return out
}

func toSources(srcs []service.PlaySourceView) []PlaySource {
	out := make([]PlaySource, 0, len(srcs))
	for _, s := range srcs {
		out = append(out, PlaySource{Id: s.Id, Name: s.Name, Episodes: toEpisodes(s.Episodes), AdFilterOk: s.AdFilterOk})
	}
	return out
}

// FilmDetail 影片详情(契约: 含 sources)。
type FilmDetail struct {
	Mid      int64        `json:"mid"`
	Name     string       `json:"name"`
	Cover    string       `json:"cover"`
	Backdrop string       `json:"backdrop,omitempty"` // 横图(16:9), 详情页 hero 背景用
	Cid      int64        `json:"cid"`
	Pid      int64        `json:"pid"`
	CName    string       `json:"cName"`
	SubTitle string       `json:"subTitle"`
	Actor    string       `json:"actor"`
	Director string       `json:"director"`
	Area     string       `json:"area"`
	Language string       `json:"language"`
	Year     int          `json:"year"`
	ClassTag string       `json:"classTag"`
	Remarks  string       `json:"remarks"`
	State    string       `json:"state"`
	DbScore  float64      `json:"dbScore"`
	// PubDate 上映日期(ISO 前缀串, 空 = 源站未提供); HotRank 豆瓣榜位(0 = 不在榜),
	// 详情页据此显示「上映日期」与「豆瓣热门 No.N」。
	// HotBoard 榜位来源榜单中文名(如 "热门电影"), 展示成「豆瓣·热门电影 No.2」;
	// 同一部片会在多个集合出现(多分类各有 No.2), 光看位次分不清是哪个榜的。
	PubDate  string       `json:"pubDate,omitempty"`
	HotRank  int          `json:"hotRank,omitempty"`
	HotBoard string       `json:"hotBoard,omitempty"`
	Content  string       `json:"content"`
	PlayFrom []string     `json:"playFrom"`
	Sources  []PlaySource `json:"sources"`
}

func ToFilmDetail(d service.FilmDetailData) FilmDetail {
	m := d.Movie
	return FilmDetail{
		Mid: m.Mid, Name: m.Name, Cover: m.Cover, Backdrop: normBackdrop(m.Backdrop), Cid: m.Cid, Pid: m.Pid, CName: m.CName,
		SubTitle: m.SubTitle, Actor: m.Actor, Director: m.Director, Area: m.Area, Language: m.Language,
		Year: m.Year, ClassTag: m.ClassTag, Remarks: m.Remarks, State: m.State, DbScore: m.DbScore,
		PubDate: m.PubDate, HotRank: m.HotRank, HotBoard: douban.BoardLabel(m.HotBoard),
		Content: m.Content, PlayFrom: []string(m.PlayFrom), Sources: toSources(d.Sources),
	}
}

// FilmDetailResp /films/{mid} 响应。
type FilmDetailResp struct {
	Detail  FilmDetail `json:"detail"`
	Related []Card     `json:"related"`
}

// PlayInfoResp /films/{mid}/play 响应(契约: currentSource 为 string sourceId)。
type PlayInfoResp struct {
	Detail         FilmDetail `json:"detail"`
	Current        Episode    `json:"current"`
	CurrentSource  string     `json:"currentSource"`
	CurrentEpisode int        `json:"currentEpisode"`
	Related        []Card     `json:"related"`
}

// HomeRow 首页区块。
type HomeRow struct {
	Nav    NavCategory `json:"nav"`
	Latest []Card      `json:"latest"`
	Hot    []Card      `json:"hot"`
}

// NavCategory 导航分类(带子分类)。
type NavCategory struct {
	Id       int64         `json:"id"`
	Pid      int64         `json:"pid"`
	Name     string        `json:"name"`
	Children []NavCategory `json:"children,omitempty"`
}

func toNav(n *entity.CategoryNode) NavCategory {
	nc := NavCategory{Id: n.Id, Pid: n.Pid, Name: n.Name}
	for _, ch := range n.Children {
		nc.Children = append(nc.Children, toNav(ch))
	}
	return nc
}

func ToNavList(nodes []*entity.CategoryNode) []NavCategory {
	out := make([]NavCategory, 0, len(nodes))
	for _, n := range nodes {
		out = append(out, toNav(n))
	}
	return out
}

// HomeResp 首页。
type HomeResp struct {
	Categories []NavCategory `json:"categories"`
	// Hot 全站跨类别热榜(「🔥 热门榜单」行) —— 与 rows[].hot(该分类热榜)口径不同。
	Hot  []Card    `json:"hot"`
	Rows []HomeRow `json:"rows"`
}

func ToHome(h service.HomeData) HomeResp {
	r := HomeResp{Categories: ToNavList(h.Categories), Hot: ToCards(h.Hot)}
	for _, row := range h.Rows {
		r.Rows = append(r.Rows, HomeRow{
			Nav:    NavCategory{Id: row.Nav.Id, Pid: row.Nav.Pid, Name: row.Nav.Name},
			Latest: ToCards(row.Latest), Hot: ToCards(row.Hot),
		})
	}
	return r
}

// ClassifyResp 分类页各榜(含标题分类)。
// ScoredCount 为该分类有评分的影片总数; 为 0 时 score 必为空, 前端不渲染高分榜分区。
type ClassifyResp struct {
	Title       *NavCategory `json:"title"`
	News        []Card       `json:"news"`
	Top         []Card       `json:"top"`
	Recent      []Card       `json:"recent"`
	Score       []Card       `json:"score"`
	ScoredCount int64        `json:"scoredCount"`
}

func ToClassify(c service.ClassifyData, title *entity.CategoryNode) ClassifyResp {
	r := ClassifyResp{
		News: ToCards(c.News), Top: ToCards(c.Top), Recent: ToCards(c.Recent),
		Score: ToCards(c.Score), ScoredCount: c.ScoredCount,
	}
	if title != nil {
		nc := toNav(title)
		r.Title = &nc
	}
	return r
}

// Filters /categories/{pid}/filters 响应(直接复用 repository.FilterOptions 的 json 形状)。
type Filters = repository.FilterOptions
